File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans =0;
    public int getDia(TreeNode node) {
        if(node == null) return -1;
        
        int l = getDia(node.left);
        int r = getDia(node.right);
        int diaTillMe = Math.max(l,r)+1;
        int maxPotentialdia = Math.max(l+r+2,diaTillMe);
        ans = Math.max(ans,maxPotentialdia);
        return diaTillMe;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null) return 0;
        getDia(root);
        return ans;
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int getDia(TreeNode* node) {
        if(!node) return -1;
        
        int l = getDia(node->left);
        int r = getDia(node->right);
        int diaTillmyNode = max(l,r)+1;
        int potentialMaxDia = max(diaTillmyNode,l+r+2);
        ans = max(ans,potentialMaxDia);
        return diaTillmyNode;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        getDia(root);
        return ans;
    }
};